<?xml version="1.0"?>
<launch>
    <arg name="camera_model"         default="BW1098OAK" />
    <arg name="camera_name"          default="oak" />
    <arg name="base_frame"           default="oak-d_frame" />
    <arg name="cam_pos_x"             default="0.0" /> <!-- Position respect to base frame (i.e. "base_link) -->
    <arg name="cam_pos_y"             default="0.0" /> <!-- Position respect to base frame (i.e. "base_link) -->
    <arg name="cam_pos_z"             default="0.0" /> <!-- Position respect to base frame (i.e. "base_link) -->
    <arg name="cam_roll"              default="1.5708" /> <!-- Orientation respect to base frame (i.e. "base_link) -->
    <arg name="cam_pitch"             default="0.0" /> <!-- Orientation respect to base frame (i.e. "base_link) -->
    <arg name="cam_yaw"               default="1.5708" /> <!-- Orientation respect to base frame (i.e. "base_link) -->
    <arg name="camera_param_uri"      default="package://oakd_pcloud/params/camera" />

    <include file="$(find depthai_bridge)/launch/urdf.launch">
        <arg name="base_frame"      value="$(arg  base_frame)"  />
        <arg name="camera_model"    value="$(arg  camera_model)"/>
        <arg name="camera_name"     value="$(arg  camera_name)" />
        <arg name="cam_pos_x"       value="$(arg  cam_pos_x)"   />
        <arg name="cam_pos_y"       value="$(arg  cam_pos_y)"   />
        <arg name="cam_pos_z"       value="$(arg  cam_pos_z)"   />
        <arg name="cam_roll"        value="$(arg  cam_roll)"    />
        <arg name="cam_pitch"       value="$(arg  cam_pitch)"   />
        <arg name="cam_yaw"         value="$(arg  cam_yaw)"     />
    </include>

    <node name="stereo_rgb_node" pkg="oakd_pcloud" type="stereo_rectified_rgb_node" output="screen" required="true">
        <param name="camera_name" value="$(arg camera_name)"/>
        <param name="camera_param_uri" value="$(arg camera_param_uri)"/>
    </node>            

    <node pkg="nodelet" type="nodelet" name="nodelet_manager"  args="manager" output="screen"/>
    
    <node pkg="nodelet" type="nodelet" name="median_filter_depth_nodelet"
        args="load oakd_pcloud/median_filter nodelet_manager">
        <remap from="depth_image" to="/stereo_rgb_node/stereo/depth"/>
        <remap from="right_image" to="/stereo_rgb_node/rectified_right/image"/>
        <remap from="right_image_info" to="/stereo_rgb_node/rectified_right/camera_info"/>
        <remap from="left_image" to="/stereo_rgb_node/rectified_left/image"/>
        <remap from="left_image_info" to="/stereo_rgb_node/rectified_left/camera_info"/>
        <remap from="median_depth" to="/stereo_rgb_node/stereo/median_depth"/>
    </node>

    <node pkg="nodelet" type="nodelet" name="depth_image_convertion_nodelet"
        args="load depth_image_proc/convert_metric nodelet_manager">
        <remap from="image_raw" to="/stereo_rgb_node/stereo/median_depth"/>
<!--         <remap from="image_raw" to="/stereo_rgb_node/stereo/depth"/>     -->
        <remap from="image" to="/stereo_rgb_node/stereo/image"/>
    </node>

    <node pkg="nodelet" type="nodelet" name="depth_image_to_pointcloud_nodelet"
        args="load depth_image_proc/point_cloud_xyz nodelet_manager">
        <remap from="camera_info" to="/stereo_rgb_node/stereo/camera_info"/>
        <remap from="image_rect" to="/stereo_rgb_node/stereo/median_depth"/>
        <remap from="points" to="/stereo_rgb_node/stereo/points"/>
    </node>

    <!-- Registration nodelet, projecting depth to RGB camera -->
    <!-- <node pkg="nodelet" type="nodelet" name="register_rgb"
            args="load depth_image_proc/register nodelet_manager">
        <remap from="rgb/camera_info"             to="/stereo_rgb_node/color/camera_info" />
        <remap from="depth/camera_info"           to="/stereo_rgb_node/stereo/camera_info" />
        <remap from="depth/image_rect"            to="/stereo_rgb_node/stereo/image" />
        <remap from="depth_registered/image_rect" to="/stereo_rgb_node/stereo/image_rect" />
    </node>

    <node pkg="nodelet" type="nodelet" name="depth_image_to_pointcloud_rgb_nodelet"
        args="load depth_image_proc/point_cloud_xyzrgb nodelet_manager">
        <remap from="depth_registered/image_rect" to="/stereo_rgb_node/stereo/image_rect"/> 
        <remap from="depth_registered/points" to="/stereo_rgb_node/stereo/points"/>
        <remap from="rgb/image_rect_color" to="/stereo_rgb_node/color/image"/>
        <remap from="rgb/camera_info" to="/stereo_rgb_node/color/camera_info"/>
    </node> -->

    <!-- Rviz configuration -->
<!--   <node type="rviz" name="rviz" pkg="rviz" args="-d $(find oakd_pcloud)/rviz/stereo_nodelet.rviz" /> -->

</launch>
